code/datums/tracking_sound.dm

```diff
@@ -0,0 +1,199 @@
+
+	start_time = REALTIMEOFDAY
+	GLOB.sound_spatial_trackers[channel] = src
+	spatial_tracker = new(max_distance, max_distance)
+	update_spatial_tracker()
+	src.sound_length = sound_length
+	if(sound_length)
+		schedule_qdel(sound_length)
+	return ..()
+
+/datum/sound_spatial_tracker/proc/schedule_qdel(length)
+	if(qdel_scheduled)
+		return
+	qdel_scheduled = QDEL_IN_STOPPABLE(src, length)
+
+/datum/sound_spatial_tracker/Destroy(force, ...)
+	source = null
+	for(var/mob/listener as anything in listeners)
+		release_listener(listener)
+	listeners.Cut()
+	if(GLOB.sound_spatial_trackers[channel] == src)
+		GLOB.sound_spatial_trackers[channel] = null
+	spatial_tracker = null
+	cells.Cut()
+	return ..()
+
+/datum/sound_spatial_tracker/process(seconds_per_tick)
+	for(var/mob/listent as anything in leavers)
+		release_listener(listent)
+	leavers.Cut()
+
+/datum/sound_spatial_tracker/proc/on_source_moved()
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, PROC_REF(update_spatial_tracker))
+
+/datum/sound_spatial_tracker/proc/update_spatial_tracker()
+	var/enter_signal = SPATIAL_GRID_CELL_ENTERED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)
+	var/leave_signal = SPATIAL_GRID_CELL_EXITED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)
+	var/list/new_and_old_cells = spatial_tracker.recalculate_cells(get_turf(source))
+	for(var/datum/spatial_grid_cell/new_cell as anything in new_and_old_cells[1])
+		RegisterSignal(new_cell, enter_signal, PROC_REF(entered_cell))
+		RegisterSignal(new_cell, leave_signal, PROC_REF(exited_cell))
+		for(var/mob/listener as anything in new_cell.client_contents)
+			link_to_listener(listener)
+		cells[new_cell] = TRUE
+	for(var/datum/spatial_grid_cell/old_cell as anything in new_and_old_cells[2])
+		UnregisterSignal(old_cell, enter_signal)
+		UnregisterSignal(old_cell, leave_signal)
+		for(var/mob/listener as anything in old_cell.client_contents)
+			release_listener(listener)
+		cells -= old_cell
+	for(var/mob/listener as anything in listeners)
+		update_listener(listener)
+
+/datum/sound_spatial_tracker/proc/link_to_listener(mob/listener)
+	if(listeners[listener])
+		return
+	RegisterSignal(listener, COMSIG_MOVABLE_MOVED, PROC_REF(on_listener_moved))
+	listeners[listener] = TRUE
+
+/datum/sound_spatial_tracker/proc/release_listener(mob/listener)
+	if(!listeners[listener])
+		return
+	listeners -= listener
+	UnregisterSignal(listener, COMSIG_MOVABLE_MOVED)
+	if(isnull(listener.client))
+		return
+	stop_sound_for(listener)
+
+/datum/sound_spatial_tracker/proc/stop_sound_for(mob/listener)
+	var/sound/null_sound = sound(null, channel = channel)
+	SEND_SOUND(listener.client, null_sound)
+
+/datum/sound_spatial_tracker/proc/update_listener(mob/listener)
+	if(isnull(listener.client))
+		return
+
+	if(get_dist(listener, source) > max_distance)
+		stop_sound_for(listener)
+		return
+	var/expected_offset = (REALTIMEOFDAY - start_time) * 0.1
+
+	var/listener_offset = listeners[listener] - 1
+	if(!listener_offset)
+		var/sound/existing_sound = null
```

Comment:
also also what happens if the offset is 0? (should we use +2 and default to 1 so it goes negative and we can tell?)